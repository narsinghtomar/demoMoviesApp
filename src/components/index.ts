import Button from './button/index.tsx';
import ButtonLoader from './buttonloader/index.tsx';
import Space from './space/index.tsx';
import InputField from './inputField/index.tsx';
import ImageView from './imageView/index.tsx';
import Header from './header/index.tsx';

export { Button, ButtonLoader, Space, InputField, ImageView, Header };
