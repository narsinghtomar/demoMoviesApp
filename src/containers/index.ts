/**
 * import
 */
import LoginPage from './login';
import MoviesListPage from './moviesList';
/**
 * export
 */
export {LoginPage, MoviesListPage};
