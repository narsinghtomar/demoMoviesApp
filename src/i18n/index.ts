import * as RNLocalize from 'react-native-localize';
import { I18nManager } from 'react-native';
import { AR, EN } from '../utils/constants';
import { I18n, LanguageTag } from 'i18n-js';

export const DefaultLanguage: LanguageTag = EN;
const i18n = new I18n();

const translationGetters = {
  en: {
    common: require('./locales/en/common.json'),
    auth: require('./locales/en/auth.json'),
  },
  ar: {
    common: require('./locales/ar/common.json'),
    auth: require('./locales/ar/auth.json'),
  },
};

export const setI18nConfig = (appLang?: LanguageTag) => {
  const isRTLBool = appLang === AR;
  i18n.fallbacks = true;
  const translations = appLang ? [appLang] : Object.keys(translationGetters);

  const fallback = {
    languageTag: appLang ?? DefaultLanguage,
    isRTL: isRTLBool,
  };

  const { languageTag } = RNLocalize?.findBestLanguageTag(translations) || fallback;
  i18n.translations = { [languageTag]: translationGetters[languageTag] };
  i18n.locale = languageTag;
  I18nManager.allowRTL(isRTLBool);
  I18nManager.forceRTL(isRTLBool);
};

// The method we'll use instead of a regular string
export function strings(key: string, params: Record<string, any> = {}): string | null {
  return key ? i18n.t(key, params) : null;
}

export function getLocale(): LanguageTag {
  return i18n.locale as LanguageTag;
}

export function getLocaleInitials(): string {
  return i18n.locale.slice(0, 2);
}

export function isRTL(): boolean {
  return I18nManager.isRTL;
}

export default i18n;
