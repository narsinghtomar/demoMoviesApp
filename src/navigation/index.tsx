import React, { useRef } from 'react'
import {
  NavigationContainer,
  NavigationContainerRef,
} from '@react-navigation/native'
import { AuthNavigator, HomeNavigator } from './StackNavigator'
import { useSelector } from 'react-redux'

interface AppNavigatorProps {
  setNavigationRef: (ref: NavigationContainerRef | null) => void
}

const AppNavigator: React.FC<AppNavigatorProps> = ({ setNavigationRef }) => {
  const globalNavigationRef = useRef<NavigationContainerRef | null>(null)

  const authReducer = useSelector((state: any) => state.authReducer)
  const { isAuthDone } = authReducer || {}

  const getNavigationStack = () => {
    if (isAuthDone) {
      return <HomeNavigator />
    } else {
      return <AuthNavigator />
    }
  }

  return (
    <NavigationContainer
      ref={(navigationRef) => {
        setNavigationRef(navigationRef)
        if (navigationRef) {
          globalNavigationRef.current = navigationRef
        }
      }}
    >
      {getNavigationStack()}
    </NavigationContainer>
  )
}

export default AppNavigator
