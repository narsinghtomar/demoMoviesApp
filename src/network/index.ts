import {
  DEFAULT_ERROR_STATUS,
  API_RESPONSE_SUCCESS,
  DEFAULT_ERROR_STRING,
} from '../utils/constants/apiCodes';
import { BASE_SERVER_URL } from '../utils/constants/apiEndpoints';

interface NetworkResponse<T> {
  status: number;
  data?: T;
  message?: string;
  error?: string;
}

interface NetworkConfig {
  url: string;
  method: string;
  headers?: Record<string, string>;
  data?: Record<string, any>;
}

/**
 * Make Fetch NetworkCall
 * @param config
 * @returns
 */
async function makeAdvanceNetworkCall<T>(config: NetworkConfig): Promise<NetworkResponse<T>> {
  const response: NetworkResponse<T> = { status: DEFAULT_ERROR_STATUS };
  const url = BASE_SERVER_URL + config.url;

  const res = await fetch(url, {
    method: config.method,
    headers: config.headers,
    body: JSON.stringify(config.data),
  });

  try {
    if (res) {
      if (res.status === API_RESPONSE_SUCCESS) {
        response.status = API_RESPONSE_SUCCESS;
        response.data = await res.json();
      } else {
        response.data = await res.json();
        response.status = res.status || DEFAULT_ERROR_STATUS;
      }
    }
  } catch (error) {
    console.log('error', error);
    response.data = { error: DEFAULT_ERROR_STRING };
    response.message = DEFAULT_ERROR_STRING;
    response.error = DEFAULT_ERROR_STRING;
  }
  return response;
}

/*
 * Make NetworkCall
 */
function makeNetworkCall<T>(config: NetworkConfig): Promise<NetworkResponse<T>> {
  return makeAdvanceNetworkCall<T>(config);
}

export { makeNetworkCall, NetworkResponse, NetworkConfig };
