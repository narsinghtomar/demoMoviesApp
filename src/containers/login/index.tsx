import React from 'react'
import {
  View,
  Text,
  TouchableWithoutFeedback,
  KeyboardAvoidingView,
  Keyboard,
  TouchableOpacity,
} from 'react-native'
import { useDispatch } from 'react-redux'
import { Button, InputField } from '../../components'
import { changeAppLang, isIOS } from '../../utils/helper'
import { loginValidationSchema } from '../../utils/validator'
import { useFormik } from 'formik'
import { strings } from '../../i18n'
import { setAuthCompleted, setUserInfo } from '../../redux/reducers/authReducer'
import styles from './styles'

interface LoginPageProps {
  navigation: any
}

const LoginPage: React.FC<LoginPageProps> = ({ navigation }) => {
  const dispatch = useDispatch() // Initialize dispatch

  // useFormik
  const {
    handleChange,
    handleSubmit,
    setFieldTouched,
    values,
    errors,
    touched,
    isValid,
  } = useFormik({
    validationSchema: loginValidationSchema(),
    initialValues: {
      email: '',
      password: '',
    },
    onSubmit: (userValues) => {
      handelUserLogin(userValues)
      Keyboard.dismiss()
    },
    validateOnChange: true,
  })

  // Determine if the form is valid based on the isValid property
  const isFormValid = isValid && Object.keys(touched).length > 0

  /**
   * handelUserLogin
   */
  const handelUserLogin = async (userValues: {
    email: string
    password: string
  }) => {
    const loginData = {
      username: userValues?.email,
      password: userValues?.password,
    }
    dispatch(setAuthCompleted())
    dispatch(setUserInfo(loginData))
  }

  /**
   * emailProps
   */
  const emailProps = () => {
    return {
      labelText: strings('auth.login.email'),
      placeholder: strings('auth.login.enterEmail'),
      value: values?.email,
      textInputName: 'email',
      errorMessage: errors?.email,
      onChangeText: (text) => {
        handleChange('email')(text)
      },
      onBlur: () => {
        setFieldTouched('email')
      },
      isErrorMsgRequired: errors?.email,
    }
  }

  /**
   * passwordProps
   */
  const passwordProps = () => {
    return {
      labelText: strings('auth.login.password'),
      placeholder: strings('auth.login.enterPassword'),
      value: values?.password,
      textInputName: 'password',
      errorMessage: errors?.password,
      secureTextEntry: true,
      onChangeText: (text) => {
        handleChange('password')(text)
      },
      onBlur: () => {
        setFieldTouched('password')
      },
      isErrorMsgRequired: errors?.password,
    }
  }

  return (
    <View style={styles.container}>
      {/*  */}
      <TouchableOpacity style={styles.langTouch} onPress={changeAppLang}>
        <Text style={styles.languageText}>{strings('common.language')}</Text>
      </TouchableOpacity>

      {/*  */}
      <Text style={styles.title}>{strings('auth.login.title')}</Text>

      {/*  */}
      <TouchableWithoutFeedback onPress={Keyboard.dismiss} accessible={false}>
        <KeyboardAvoidingView
          style={styles.avoidingView}
          behavior={isIOS() ? 'padding' : 'height'}
          keyboardVerticalOffset={5}
        >
          {/* Input */}
          {InputField(emailProps())}
          {InputField(passwordProps())}

          {/* Submit */}
          <Button
            buttonStyle={styles.buttonStyle}
            text={strings('auth.login.login')}
            textStyle={styles.btnTxtStyle}
            onPress={handleSubmit}
            disabled={!isFormValid}
          />
          {/*  */}
        </KeyboardAvoidingView>
      </TouchableWithoutFeedback>
    </View>
  )
}

export default LoginPage
