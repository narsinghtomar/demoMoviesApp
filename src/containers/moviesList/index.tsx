import React, { useCallback, useEffect, useRef, useState } from 'react'
import { View, FlatList, ActivityIndicator, RefreshControl } from 'react-native'
import { useDispatch, useSelector } from 'react-redux'
import { fetchPopularMoviesRequest } from '../../redux/reducers/moviesListReducer'
import { isRTL, strings } from '../../i18n'
import { generateHeaders } from '../../utils/constants/apiHeaders'
import MoviesListItem from './moviesListItem'
import styles from './styles'
import { Header, Space } from '../../components'
import { AR, EN_US } from '../../utils/constants'

const MoviesListPage = ({ navigation }) => {
  const flatlistRef = useRef<FlatList>(null)
  const dispatch = useDispatch()
  const [page, setPage] = useState<number>(1)

  // useSelector
  const moviesListReducer = useSelector((state) => state.moviesListReducer)
  const {
    popularMoviesList = [],
    isMoviesPaginating,
    isMoviesPageFinished,
    isPopularMoviesLoading,
  } = moviesListReducer || {}

  useEffect(() => {
    handleFetchPopularMovies(1)
  }, [])

  /**
   * API Calls
   * Fetch Popular Movies
   */
  const handleFetchPopularMovies = useCallback(
    async (page: number, isRefresh = false) => {
      const bodyData = {
        page: page,
        language: isRTL() ? AR : EN_US,
      }
      const headers = await generateHeaders()
      dispatch(
        fetchPopularMoviesRequest({
          bodyData: bodyData,
          headers: headers,
          isRefresh: isRefresh,
        })
      )
    },
    [dispatch]
  )

  /**
   * callPullToRefresh
   */
  const callPullToRefresh = () => {
    setPage(1)
    handleFetchPopularMovies(1, true)
  }

  const onItemPress = useCallback((_item: any) => {}, [])

  const renderItem = ({ item, index }) => {
    return (
      <MoviesListItem
        item={item}
        index={index}
        onItemPress={onItemPress}
        key={index?.toString()}
      />
    )
  }

  /**
   * method for handling pagination when user reach to end of the list
   */
  const handleLoadMoreData = useCallback(
    (info) => {
      if (
        info?.distanceFromEnd > 0 &&
        !isMoviesPaginating &&
        !isMoviesPageFinished
      ) {
        const nextPage = page + 1
        setPage(nextPage)
        handleFetchPopularMovies(nextPage)
      }
    },
    [setPage, page]
  )

  /**
   * keyExtractor
   */
  const keyExtractor = useCallback((item) => {
    return item?.id
  }, [])

  /**
   * Returns footer section with ActivityIndicator
   */
  const listFooter = () => {
    if (!isMoviesPageFinished) {
      return (
        <ActivityIndicator
          size={'large'}
          color={styles.activityIndicatorColor}
        />
      )
    } else {
      return <Space height={100} />
    }
  }

  return (
    <View style={styles.container}>
      <Header
        headerTitle={strings('common.movies.popularMovies')}
        navigation={navigation}
      />
      <View style={styles.flatListContainer}>
        <FlatList
          ref={flatlistRef}
          numColumns={2}
          data={popularMoviesList}
          renderItem={renderItem}
          keyExtractor={keyExtractor}
          showsVerticalScrollIndicator={false}
          contentContainerStyle={styles.scrollContent}
          onEndReached={handleLoadMoreData}
          onEndReachedThreshold={0.5}
          ListFooterComponent={listFooter}
          removeClippedSubviews={true}
          initialNumToRender={10}
          maxToRenderPerBatch={5}
          windowSize={10}
          refreshControl={
            <RefreshControl
              refreshing={isPopularMoviesLoading}
              onRefresh={() => callPullToRefresh()}
            />
          }
        />
      </View>
    </View>
  )
}

export default MoviesListPage
