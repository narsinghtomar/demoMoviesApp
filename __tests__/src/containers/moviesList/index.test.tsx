/**
 * MoviesListPage
 */
import 'react-native'
import React from 'react'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import { MoviesListPage } from '../../../../src/containers'
import configureStore from 'redux-mock-store'

it('MoviesListPage component renders correctly', async () => {
  const initialState = {}
  const mockStore = configureStore()
  let store
  jest.useFakeTimers()

  store = mockStore(initialState)
  const tree = await renderer
    .create(
      <Provider store={store}>
        <MoviesListPage />
      </Provider>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
