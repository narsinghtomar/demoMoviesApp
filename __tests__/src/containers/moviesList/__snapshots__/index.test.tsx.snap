// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesListPage component renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    style={
      [
        {
          "backgroundColor": "#EB8D0C",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        [
          {
            "alignContent": "center",
            "alignItems": "center",
            "alignSelf": "center",
            "flexDirection": "row",
            "justifyContent": "center",
            "paddingBottom": 8,
            "paddingTop": 0,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          [
            {
              "alignSelf": "center",
              "color": "#FFFFFF",
              "flex": 1,
              "fontSize": 18,
              "fontWeight": "bold",
              "justifyContent": "center",
              "marginLeft": 0,
              "paddingRight": 16,
              "textAlign": "center",
            },
            {},
          ]
        }
      >
        [missing "en.common.movies.popularMovies" translation]
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "bottom": 2,
            "opacity": 1,
            "padding": 5,
            "position": "absolute",
            "right": 10,
          }
        }
      >
        <Image
          resizeMode="contain"
          source={
            {
              "testUri": "../../../src/assets/images/logout/logout.png",
            }
          }
          style={
            {
              "height": 28,
              "tintColor": "#FFFFFF",
              "width": 28,
            }
          }
        />
      </View>
    </View>
  </RCTSafeAreaView>
  <View
    style={
      {
        "marginTop": 10,
      }
    }
  >
    <RCTScrollView
      ListFooterComponent={[Function]}
      contentContainerStyle={
        {
          "paddingVertical": 5,
        }
      }
      data={[]}
      getItem={[Function]}
      getItemCount={[Function]}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={5}
      onContentSizeChange={[Function]}
      onEndReached={[Function]}
      onEndReachedThreshold={0.5}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      refreshControl={
        <RefreshControlMock
          onRefresh={[Function]}
        />
      }
      removeClippedSubviews={true}
      renderItem={[Function]}
      scrollEventThrottle={0.0001}
      showsVerticalScrollIndicator={false}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
      windowSize={10}
    >
      <RCTRefreshControl />
      <View>
        <View
          onLayout={[Function]}
        >
          <ActivityIndicator
            size="large"
          />
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
